receiver/dynamic/receiver.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dynamic
+
+import (
+	"errors"
+	"fmt"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/carbonreceiver"
+	"github.com/open-telemetry/opentelemetry-collector/component"
+	"github.com/open-telemetry/opentelemetry-collector/consumer"
+	"github.com/open-telemetry/opentelemetry-collector/receiver"
+	"go.uber.org/zap"
+	"sync"
+)
+
+var (
+	errNilNextConsumer = errors.New("nil nextConsumer")
+	errAlreadyStarted  = errors.New("already started")
+	errAlreadyStopped  = errors.New("already stopped")
+	factories          = buildFactoryMap(&carbonreceiver.Factory{})
+)
+
+func buildFactoryMap(factories ...receiver.Factory) map[string]receiver.Factory {
+	ret, err := receiver.Build(factories...)
+	if err != nil {
+		panic(fmt.Sprintf("failed to build factory map for dynamic receiver: %s", err))
+	}
+	return ret
+}
+
+var _ receiver.MetricsReceiver = (*dynamicReceiver)(nil)
+
+// dynamicReceiver implements consumer.MetricsConsumer.
+type dynamicReceiver struct {
+	sync.Mutex
+	nextConsumer consumer.MetricsConsumer
+	logger       *zap.Logger
+
+	startOnce sync.Once
+	stopOnce  sync.Once
+}
+
+// New creates the CollectD receiver with the given parameters.
```

Comment:
```suggestion
// New creates the ?? receiver with the given parameters.
```
I assume this is a typo?